### ISSUE TITLE
Able to pass arguments to function which do not match typing.

### ISSUE THREAD

**Issue by isuda** [Microsoft/TypeScript#14159]
**TypeScript Version:**  2.1.6

**Code**

```ts

interface Data {
  something: string;
  somethingElse: number;
}

function test(arg: Partial<Data>) { }
function test2(arg: { [key: string]: any }) { }
function test3(arg: { [key: string]: string }) { }

test({ something: "true" }); // OK
test(() => false); // OK
test(false); // OK
test(true); // OK
test([]); // OK

test2({ something: "true" }); // OK
test2(() => false); // OK
test2(false); // Argument of type false not assignable to...
test2(true); // Argument of type true not assignable to...
test2([]); // OK

test3({ something: "true" }); // OK
test3(() => false); // Argument of type () => boolean not assignable to...
test3(false); // Argument of type false not assignable to...
test3(true); // Argument of type true not assignable to...
test3([]); // Argument of type undefined[] not assignable to...


```

**Expected behavior:**
Compiler should throw an error when assigning parameters which do not match the interface or parameter type in a function call.

**Actual behavior:**
test3 has expected behaviour. test2 and test1 have unexpected behaviour on most types of invalid arguments.

**Comment by arciisine** [comment]
The playground (http://bit.ly/2lfXje9) seems to agree with your assessment.  Perhaps your tsconfig is off?

**Comment by isuda** [comment]
Yeah, correct. What I am saying is that in the first and 2nd tests, there are false positives, compared to test3, which is correctly reporting the errors.

**Comment by arciisine** [comment]
I think the way open types are handled, `[]` or `() => false` is assignable to `{[key:string]: any}`.  The `Partial<Data>` being assignable to `true|false` does seem a little weird though, but that may be me not understanding exactly how Partial works.
